ocaml/typing/ctype.ml

```diff
@@ -2091,6 +2110,55 @@ let tvariant_not_immediate row =
       | _ -> false)
     (row_fields row)
 
+
+(* Jkinds *)
+
+let is_datatype_decl (k : type_decl_kind) =
+  match k with
+  | Type_record _ | Type_variant _ | Type_open -> true
+  | Type_abstract _ -> false
+
+let find_type_opt p env =
+  try Some (Env.find_type p env) with Not_found -> None
+
+let rec jkind_of_decl_unapplied env (decl : type_declaration) : jkind option =
+  (* FIXME jbachurski: Shouldn't we look at type_variance and type_separability here? *)
+  match decl.type_arity with
+  | 0 -> Some decl.type_jkind
+  | _ when is_datatype_decl decl.type_kind ->
+    Some (Arrow
+    { args = List.map (type_jkind env) decl.type_params
+    ; result = decl.type_jkind })
+  | _ -> None
+
+and type_jkind_for_app (app_jkind : jkind) app_arity tys : jkind option =
```

Comment:
I think this would be clearer if `type_jkind_for_app_decl` was replaced by a general `jkind_of_decl_applied`, and the logic for the applied case was all in there directly. `estimate_type_jkind` can then check that the `jkind` is in fact a `type_jkind` itself. (Or possibly it should become `estimate_jkind` and the specialisation to type jkinds pushed even further out).

This function can then be simplified to only handle the arity 0 case for a list of types.